### PR TITLE
Fix deploy bugs, other bugs and exception handling and logging

### PR DESCRIPTION
**Pull request by DavidMinarsch**
## Proposed changes

This PR fixes some bugs noticed during deployment.

## Fixes

- #1919
- #1703 

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [ ] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [ ] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that code coverage does not decrease.
- [ ] I have checked that the documentation about the `aea cli` tool works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

-

### CODE REVIEW

File: aea/helpers/async_utils.py [hunk 1]
```diff
@@ -543,7 +543,7 @@ async def _run_wrapper(self) -> None:
         """Wrap run() method."""
         if not self._completed_event or not self._loop:  # pragma: nocover
             raise ValueError("Start was not called!")
-
+        self._is_running = True
```
Comment (DavidMinarsch) [inline aea/helpers/async_utils.py:546]: This was missing, revert and run reverted associated test to see warning @marcofavorito

**Review by marcofavorito — APPROVED** [review]
LGTM

File: aea/helpers/async_utils.py (hunk 1 above)
Comment (DavidMinarsch) [reply aea/helpers/async_utils.py:546]: Could set `_is_running=False` after line 551 @solarw ?

File: aea/aea.py [hunk 2]
```diff
@@ -434,7 +434,6 @@ def teardown(self) -> None:
 
         :return: None
         """
-        self.logger.debug("Calling teardown method...")
```
Comment (DavidMinarsch) [inline aea/aea.py:437]: For symmetry removed here and added in runtime

File: aea/cli/run.py [hunk 3]
```diff
@@ -129,7 +128,5 @@ def _build_aea(
         )
         aea = builder.build(connection_ids=connection_ids)
         return aea
-    except AEAPackageLoadingError as e:  # pragma: nocover
-        raise click.ClickException("Package loading error: {}".format(str(e)))
```
Comment (DavidMinarsch) [inline aea/cli/run.py:133]: Modification done higher up

File: aea/components/loader.py [hunk 4]
```diff
@@ -133,7 +133,7 @@ def get_new_error_message_with_package_found() -> str:
         new_message = get_new_error_message_with_package_found()
 
     raise AEAPackageLoadingError(
-        "An error occurred while loading {} {}: No module named {}; {}".format(
+        "Package loading error: An error occurred while loading {} {}: No module named {}; {}".format(
```
Comment (DavidMinarsch) [inline aea/components/loader.py:136]: See above

File: aea/components/loader.py [hunk 5]
```diff
@@ -150,8 +150,9 @@ def _handle_error_while_loading_component_generic_error(
 
     :raises Exception: the same exception, but prepending an informative message.
     """
+    e_str = parse_exception(e)
     raise Exception(
```
Comment (DavidMinarsch) [inline aea/components/loader.py:154]: Also make it a specific exception?

File: aea/helpers/async_utils.py [hunk 6]
```diff
@@ -543,12 +543,13 @@ async def _run_wrapper(self) -> None:
...
         try:
             with suppress(asyncio.CancelledError):
                 return await self.run()
         finally:
             self._loop.call_soon_threadsafe(self._completed_event.set)
+            self._is_running = False
```
Comment (DavidMinarsch) [inline aea/helpers/async_utils.py:552]: As discussed with @solarw

File: aea/registries/base.py [hunk 7]
```diff
@@ -471,7 +472,11 @@ def setup(self) -> None:
                         item.name, item.skill_id
                     )
                 )
-                item.setup()
+                try:
+                    item.setup()
+                except Exception as e:
+                    e_str = parse_exception(e)
+                    raise AEASetupError(e_str)
```
Comment (DavidMinarsch) [inline aea/registries/base.py:479]: Was missing before!

File: aea/runtime.py [hunk 8]
```diff
@@ -193,6 +193,7 @@ def _teardown(self) -> None:
         if self._decision_maker is not None:  # pragma: nocover
             self.decision_maker.stop()
         self.task_manager.stop()
+        self.logger.debug("[{}]: Calling teardown method...".format(self._agent.name))
```
Comment (DavidMinarsch) [inline aea/runtime.py:196]: See above

File: aea/aea.py [hunk 9]
```diff
@@ -338,7 +338,7 @@ def handle_envelope(self, envelope: Envelope) -> None:
             return
 
         for handler in handlers:
-            handler.handle(msg)
+            handler.handle_wrapper(msg)
```
Comment (DavidMinarsch) [inline aea/aea.py:341]: we wrap to be able to properly catch exceptions

File: aea/aea.py [hunk 10]
```diff
@@ -402,10 +402,14 @@ def exception_handler(self, exception: Exception, function: Callable) -> bool:
...
-        def log_exception(e, fn):
-            self.logger.exception(f"<{e}> raised during `{fn}`")
+        def log_exception(e, fn, is_debug: bool = False):
+            if is_debug:
+                self.logger.debug(f"<{e}> raised during `{fn}`")
+            else:
+                self.logger.exception(f"<{e}> raised during `{fn}`")
 
         if self._skills_exception_policy == ExceptionPolicyEnum.propagate:
+            log_exception(exception, function, is_debug=True)
```
Comment (DavidMinarsch) [inline aea/aea.py:412]: need to only log on exception level for two policies!

File: aea/agent_loop.py [hunk 11]
```diff
@@ -179,9 +179,6 @@ def _periodic_task_exception_callback(
 
         :return: None
         """
-        self.logger.exception(
```
Comment (DavidMinarsch) [inline aea/agent_loop.py:182]: This polluted the logs

File: aea/components/loader.py [hunk 12]
```diff
@@ -59,6 +59,8 @@ def load_component_from_config(  # type: ignore
     component_class = component_type_to_class(component_type)
     try:
         return component_class.from_config(*args, configuration=configuration, **kwargs)  # type: ignore
+    except AEAInstantiationException as e:
+        raise e
```
Comment (DavidMinarsch) [inline aea/components/loader.py:63]: just forward

File: aea/components/loader.py [hunk 13]
```diff
@@ -152,6 +154,6 @@ def _handle_error_while_loading_component_generic_error(
     """
     raise Exception(
         "An error occurred while loading {} {}: {}".format(
-            str(configuration.component_type), configuration.public_id, str(e)
+            str(configuration.component_type), configuration.public_id, e
         )
-    ) from e
+    )
```
Comment (DavidMinarsch) [inline aea/components/loader.py:159]: will revert

File: aea/connections/base.py [hunk 14]
```diff
@@ -268,12 +268,19 @@ def from_config(
...
+                identity=identity,
+                crypto_store=crypto_store,
+                **kwargs,
+            )
+        except Exception as e:
+            e_str = parse_exception(e)
+            raise AEAInstantiationException(
+                f"An error occured during instantiation of connection {configuration.public_id}/{configuration.class_name}:\n{e_str}"
+            )
+        return connection
```
Comment (DavidMinarsch) [inline aea/connections/base.py:283]: We now wrap all instantiations to be able to catch the exception later

File: aea/registries/base.py [hunk 15]
```diff
@@ -471,7 +472,12 @@ def setup(self) -> None:
...
                     )
                 )
-                item.setup()
+                try:
+                    item.setup()
+                except Exception as e:
+                    e_str = parse_exception(e)
+                    e_str = f"An error occurred while setting up item {item.skill_id}/{type(item).__name__}:\n{e_str}"
+                    raise AEASetupError(e_str)
             else:
```
Comment (DavidMinarsch) [inline aea/registries/base.py:481]: Specific error for setup; currently not caught anywhere but good to have!

File: aea/registries/base.py [hunk 16]
```diff
@@ -487,14 +493,18 @@ def teardown(self) -> None:
...
                     item.teardown()
                 except Exception as e:  # pragma: nocover # pylint: disable=broad-except
-                    self.logger.warning(
-                        "An error occurred while tearing down item {}/{}: {}".format(
-                            item.skill_id, type(item).__name__, str(e)
-                        )
-                    )
+                    e_str = parse_exception(e)
+                    e_str = f"An error occurred while tearing down item {item.skill_id}/{type(item).__name__}:\n{str(e_str)}"
+                    e = AEATeardownError(e_str)
```
Comment (DavidMinarsch) [inline aea/registries/base.py:506]: for consistency

File: aea/skills/base.py [hunk 17]
```diff
@@ -366,7 +373,13 @@ def is_done(self) -> bool:  # pylint: disable=no-self-use
...
     def act_wrapper(self) -> None:
         """Wrap the call of the action. This method must be called only by the framework."""
-        self.act()
+        try:
+            self.act()
+        except Exception as e:
+            e_str = parse_exception(e)
+            raise AEAActException(
+                f"An error occured during act of behaviour {self.context.skill_id}/{type(self).__name__}:\n{e_str}"
+            )
```
Comment (DavidMinarsch) [inline aea/skills/base.py:382]: Act exception introduced, currently not explicitly handled, but useful!

File: aea/skills/base.py [hunk 18]
```diff
@@ -432,12 +445,18 @@ def parse_module(  # pylint: disable=arguments-differ
...
+                        name=behaviour_id,
+                        configuration=behaviour_config,
+                        skill_context=skill_context,
+                        **dict(behaviour_config.args),
+                    )
+                except Exception as e:
+                    e_str = parse_exception(e)
+                    raise AEAInstantiationException(
+                        f"An error occured during instantiation of behaviour {skill_context.skill_id}/{behaviour_config.class_name}:\n{e_str}"
+                    )
```
Comment (DavidMinarsch) [inline aea/skills/base.py:459]: see above

**Review by marcofavorito — APPROVED** [review]
LGTM

**Review by solarw — APPROVED** [review]
LGTM

**Pull request merged**